### PR TITLE
hashmap: Store hashes as usize internally

### LINKED ISSUE #36567: Use `usize` instead of `u64` for hashes in HashMap
Just a note on the current implementation which stores the hash values per bucket as u64. We can't use more than `usize`'s bits of a hash to select a bucket anyway, so we only need to store that part in the table. This would be an improvement on 32-bit platforms to make the hashmap's data structures smaller, hopefully improving cache utilization during insert and lookup.

### PR DESCRIPTION
**Pull request by bluss**
We can't use more than usize's bits of a hash to select a bucket anyway,
so we only need to store that part in the table. This should be an
improvement for the size of the data structure on 32-bit platforms.
Smaller data means better cache utilization and hopefully better
performance.

Fixes #36567

### CODE REVIEW

**Comment by bluss** [comment]
Needs benchmarks and testing on 32-bit.

**Comment by arthurprs** [comment]
I was doing some related benchmarks and decided to give this a go

```
➜  hashmap2 git:(layout) ✗ cargo benchcmp hhkkvv_u64:: hhkkvv_usize:: x86.txt  
 name                        hhkkvv_u64:: ns/iter  hhkkvv_usize:: ns/iter  diff ns/iter   diff % 
 grow_10_000                 1,294,797             1,166,500                   -128,297   -9.91% 
 grow_big_value_10_000       4,245,360             3,852,706                   -392,654   -9.25% 
 grow_fnv_10_000             434,746               383,631                      -51,115  -11.76% 
 insert_100                  5,404                 4,615                           -789  -14.60% 
 insert_1000                 47,765                47,701                           -64   -0.13% 
 insert_100_000              6,935,995             7,091,894                    155,899    2.25% 
 insert_10_000               548,813               521,256                      -27,557   -5.02% 
 insert_int_bigvalue_10_000  1,447,826             1,377,720                    -70,106   -4.84% 
 insert_str_10_000           630,196               592,972                      -37,224   -5.91% 
 insert_string_10_000        1,385,059             1,370,400                    -14,659   -1.06% 
 iter_keys_10_000            56,144                60,206                         4,062    7.23% 
 iter_keys_big_value_10_000  67,847                74,115                         6,268    9.24% 
 iter_values_10_000          61,340                85,706                        24,366   39.72% 
 iterate_10_000              62,875                80,937                        18,062   28.73% 
 lookup_100_000              334,371               311,320                      -23,051   -6.89% 
 lookup_100_000_bigvalue     321,824               305,257                      -16,567   -5.15% 
 lookup_10_000               272,064               242,068                      -29,996  -11.03% 
 lookup_10_000_bigvalue      287,571               265,069                      -22,502   -7.82% 
 lookup_10_000_exist         252,040               243,920                       -8,120   -3.22% 
 lookup_10_000_noexist       271,513               274,763                        3,250    1.20% 
 lookup_1_000_000            263,131               255,507                       -7,624   -2.90% 
 lookup_1_000_000_bigvalue   274,404               259,742                      -14,662   -5.34% 
 merge_shuffle               1,688,701             1,571,550                   -117,151   -6.94% 
 merge_simple                46,925,438            38,608,876                -8,316,562  -17.72% 
 new                         9                     9                                  0    0.00% 
 with_capacity_10e5          2,545                 1,362                         -1,183  -46.48%
```

what I can't understand for the life of me is how iteration becomes slower. My benchmarks must be busted... I just can't see why.

```
        #[bench]
        fn iterate_10_000(b: &mut Bencher) {
            let c = 10_000usize;
            let mut map = $hashmap::with_capacity(c);
            for x in 0..c {
                map.insert(x, x);
            }
            b.iter(|| {
                map.iter().map(|(k, v)| *k + *v).sum::<usize>()
            });
        }

        #[bench]
        fn iter_keys_10_000(b: &mut Bencher) {
            let c = 10_000usize;
            let mut map = $hashmap::with_capacity(c);
            for x in 0..c {
                map.insert(x, x);
            }
            b.iter(|| {
                map.keys().sum::<usize>()
            });
        }

        #[bench]
        fn iter_values_10_000(b: &mut Bencher) {
            let c = 10_000usize;
            let mut map = $hashmap::with_capacity(c);
            for x in 0..c {
                map.insert(x, x);
            }
            b.iter(|| {
                map.values().sum::<usize>()
            });
        }
```

**Comment by alexcrichton** [comment]
@bluss we discussed this briefly at libs triage today and were curious, do you have some representative numbers as well about this change? (or @arthurprs do you have some?) Would be useful to see the comparisons!

I was personally a little worried about losing 32 bits of a 64-bit hash on 32-bit platforms, but so long as we generally recommend a uniform distribution of bits I think it'll work out.

File: src/libstd/collections/hash/table.rs [hunk 1]
```diff
@@ -137,15 +148,28 @@ pub struct GapThenFull<K, V, M> {
...
+    pub fn inspect(&self) -> HashUint {
         self.hash
     }
+
+    #[inline(always)]
+    pub fn new(hash: u64) -> Self {
+        // We need to avoid 0 in order to prevent collisions with
+        // EMPTY_HASH. We can maintain our precious uniform distribution
+        // of initial indexes by unconditionally setting the MSB,
+        // effectively reducing 64-bits hashes to 63 bits.
```
Comment (arthurprs) [inline src/libstd/collections/hash/table.rs]: This comment needs updating because of the 32 -> 31?

**Comment by arthurprs** [comment]
Here are some benchmarks. Keep in mind that most of these are `<usize, usize>` which is the best case for this change (memory footprint is 25% smaller).

What I don't get is why in the hhkkvv layout the iteration becomes slower, I'll try to dig something from the assembly.

```
➜  hashmap2 git:(layout) ✗ cargo benchcmp hhkkvv_u64:: hhkkvv:: x86.txt
 name                            hhkkvv_u64:: ns/iter  hhkkvv:: ns/iter  diff ns/iter   diff % 
 grow_10_000                     1,199,628             1,207,945                8,317    0.69% 
 grow_big_value_10_000           3,878,615             3,895,867               17,252    0.44% 
 grow_fnv_10_000                 437,103               378,952                -58,151  -13.30% 
 insert_100                      4,920                 5,155                      235    4.78% 
 insert_1000                     47,755                46,263                  -1,492   -3.12% 
 insert_100_000                  6,879,818             7,213,230              333,412    4.85% 
 insert_10_000                   543,583               529,339                -14,244   -2.62% 
 insert_int_bigvalue_10_000      1,447,643             1,447,904                  261    0.02% 
 insert_str_10_000               629,221               599,453                -29,768   -4.73% 
 insert_string_10_000            1,391,713             1,382,703               -9,010   -0.65% 
 iter_keys_100_000               356,035               384,421                 28,386    7.97% 
 iter_keys_big_value_100_000     368,610               388,422                 19,812    5.37% 
 iter_values_100_000             394,876               551,698                156,822   39.71% 
 iterate_100_000                 394,146               561,685                167,539   42.51% 
 lookup_100_000                  330,594               311,451                -19,143   -5.79% 
 lookup_100_000_bigvalue         316,673               304,491                -12,182   -3.85% 
 lookup_10_000                   265,984               251,562                -14,422   -5.42% 
 lookup_10_000_bigvalue          283,698               266,500                -17,198   -6.06% 
 lookup_10_000_exist             248,429               247,662                   -767   -0.31% 
 lookup_10_000_noexist           271,572               274,507                  2,935    1.08% 
 lookup_1_000_000                260,549               253,014                 -7,535   -2.89% 
 lookup_1_000_000_bigvalue       266,807               263,100                 -3,707   -1.39% 
 lookup_1_000_000_bigvalue_unif  915,335               1,007,780               92,445   10.10% 
 lookup_1_000_000_unif           889,155               1,024,102              134,947   15.18% 
 merge_shuffle                   1,628,946             1,597,806              -31,140   -1.91% 
 merge_simple                    62,462,585            43,020,688         -19,441,897  -31.13% 
 new                             9                     9                            0    0.00% 
 with_capacity_10e5              2,537                 1,371                   -1,166  -45.96% 
➜  hashmap2 git:(layout) ✗ cargo benchcmp hhkvkv_u64:: hhkvkv:: x86.txt
 name                            hhkvkv_u64:: ns/iter  hhkvkv:: ns/iter  diff ns/iter   diff % 
 grow_10_000                     1,154,596             1,055,829              -98,767   -8.55% 
 grow_big_value_10_000           3,426,091             3,392,353              -33,738   -0.98% 
 grow_fnv_10_000                 417,095               359,148                -57,947  -13.89% 
 insert_100                      4,649                 4,546                     -103   -2.22% 
 insert_1000                     46,619                44,843                  -1,776   -3.81% 
 insert_100_000                  6,561,571             6,082,426             -479,145   -7.30% 
 insert_10_000                   545,366               507,511                -37,855   -6.94% 
 insert_int_bigvalue_10_000      1,163,796             1,147,447              -16,349   -1.40% 
 insert_str_10_000               591,783               579,419                -12,364   -2.09% 
 insert_string_10_000            1,359,738             1,358,532               -1,206   -0.09% 
 iter_keys_100_000               357,701               335,513                -22,188   -6.20% 
 iter_keys_big_value_100_000     580,259               519,009                -61,250  -10.56% 
 iter_values_100_000             358,010               340,095                -17,915   -5.00% 
 iterate_100_000                 363,865               343,013                -20,852   -5.73% 
 lookup_100_000                  318,030               293,466                -24,564   -7.72% 
 lookup_100_000_bigvalue         328,763               307,607                -21,156   -6.44% 
 lookup_10_000                   268,099               249,152                -18,947   -7.07% 
 lookup_10_000_bigvalue          273,723               257,195                -16,528   -6.04% 
 lookup_10_000_exist             248,437               241,170                 -7,267   -2.93% 
 lookup_10_000_noexist           274,965               272,160                 -2,805   -1.02% 
 lookup_1_000_000                258,771               249,529                 -9,242   -3.57% 
 lookup_1_000_000_bigvalue       270,629               256,560                -14,069   -5.20% 
 lookup_1_000_000_bigvalue_unif  884,262               888,794                  4,532    0.51% 
 lookup_1_000_000_unif           851,032               750,496               -100,536  -11.81% 
 merge_shuffle                   1,518,106             1,405,754             -112,352   -7.40% 
 merge_simple                    37,792,105            30,285,016          -7,507,089  -19.86% 
 new                             9                     9                            0    0.00% 
 with_capacity_10e5              2,513                 1,326                   -1,187  -47.23% 
➜  hashmap2 git:(layout) ✗
```

**Comment by bluss** [comment]
I wanted to know how it influences bootstrap time on both 64-bit and 32-bit, it would be a good metric. I don't have the resources to do the comparison, so I left this here to ask if someone wanted to try it.

Ah, the benchmarks are for kkvv layout (current) and kvkv (your pr)

**Comment by arthurprs** [comment]
I looked at the disassembly but I can't see anything that would make iteration that much slower. From an algorithm POV it makes no sense to me.

**Comment by alexcrichton** [comment]
@arthurprs those numbers seem to be mostly related to performance, but isn't the theory behind this change that it mostly saves memory?

**Comment by arthurprs** [comment]
@alexcrichton sure, I just happened to have these around due to the other PR. And the decrease in cache pressure shows up in these.

As for memory benefits it's easy to calculate as a function of sizeof(K) and sizeof(V) as it doesn't depend on other factors like padding etc.

| sz(K)+sz(V)  |  4      |  8     |  16     |  24     |  32   |  64     |  128    |  256   | 
|--------------|---------|--------|---------|---------|-------|---------|---------|--------| 
| (x+4)/(x+8)  |  0.667  |  0.75  |  0.833  |  0.875  |  0.9  |  0.944  |  0.971  |  0.985 |

**Comment by bluss** [comment]
@alexcrichton Smaller data means better performance; more hashes fit into a cache line during lookup. So performance was my main thought.

**Comment by arthurprs** [comment]
Dealing with so many benchmarks is tricky. I found a problem so I'm repeating the benchmarks, will post fixed and expanded updates soon.

**Comment by arthurprs** [comment]
Here's the updated results. And the gist with even more comparisons and better visualization https://gist.github.com/arthurprs/9f28847dceee86bd5cfffcd30d9cd6cc

```
➜  hashmap2 git:(layout_usize) ✗ cargo benchcmp hhkkvv_u64:: hhkkvv_usz:: x86.txt
 name                            hhkkvv_u64:: ns/iter  hhkkvv_usz:: ns/iter  diff ns/iter   diff % 
 grow_100_000                    1,288,426             1,347,922                   59,496    4.62% 
 grow_10_000                     1,302,671             1,351,398                   48,727    3.74% 
 grow_big_value_100_000          38,371,161            37,769,065                -602,096   -1.57% 
 grow_big_value_10_000           4,054,619             4,437,890                  383,271    9.45% 
 grow_fnv_10_000                 433,147               380,392                    -52,755  -12.18% 
 insert_100                      5,215                 4,970                         -245   -4.70% 
 insert_1000                     48,039                44,946                      -3,093   -6.44% 
 insert_100_000                  6,843,047             7,122,700                  279,653    4.09% 
 insert_10_000                   532,906               530,855                     -2,051   -0.38% 
 insert_1_000_000                124,166,274           124,684,470                518,196    0.42% 
 insert_int_bigvalue_10_000      1,440,912             1,390,695                  -50,217   -3.49% 
 insert_str_10_000               626,603               595,789                    -30,814   -4.92% 
 insert_string_10_000            1,370,791             1,371,131                      340    0.02% 
 iter_keys_100_000               371,087               383,975                     12,888    3.47% 
 iter_keys_1_000_000             8,448,411             8,579,116                  130,705    1.55% 
 iter_keys_big_value_100_000     352,423               393,528                     41,105   11.66% 
 iter_keys_big_value_1_000_000   8,386,089             8,615,999                  229,910    2.74% 
 iter_values_100_000             392,468               545,712                    153,244   39.05% 
 iter_values_1_000_000           8,797,970             11,673,560               2,875,590   32.68% 
 iterate_100_000                 394,909               552,853                    157,944   40.00% 
 iterate_1_000_000               8,778,111             11,821,943               3,043,832   34.68% 
 lookup_100_000                  328,320               313,771                    -14,549   -4.43% 
 lookup_100_000_bigvalue         319,496               316,429                     -3,067   -0.96% 
 lookup_10_000                   265,687               252,238                    -13,449   -5.06% 
 lookup_10_000_bigvalue          281,491               261,869                    -19,622   -6.97% 
 lookup_10_000_exist             257,790               246,435                    -11,355   -4.40% 
 lookup_10_000_noexist           273,700               273,483                       -217   -0.08% 
 lookup_1_000_000                257,266               251,604                     -5,662   -2.20% 
 lookup_1_000_000_bigvalue       264,313               261,944                     -2,369   -0.90% 
 lookup_1_000_000_bigvalue_unif  684,819               740,230                     55,411    8.09% 
 lookup_1_000_000_unif           598,199               619,402                     21,203    3.54% 
 merge_shuffle                   1,604,293             1,636,942                   32,649    2.04% 
 merge_simple                    61,494,395            42,660,176             -18,834,219  -30.63% 
 new                             9                     9                                0    0.00% 
 with_capacity_10e5              2,537                 1,308                       -1,229  -48.44% 
```

```
➜  hashmap2 git:(layout_usize) ✗ cargo benchcmp hhkvkv_u64:: hhkvkv_usz:: x86.txt 
 name                            hhkvkv_u64:: ns/iter  hhkvkv_usz:: ns/iter  diff ns/iter   diff % 
 grow_100_000                    1,162,181             1,138,593                  -23,588   -2.03% 
 grow_10_000                     1,158,647             1,084,507                  -74,140   -6.40% 
 grow_big_value_100_000          37,001,012            36,665,075                -335,937   -0.91% 
 grow_big_value_10_000           3,437,647             3,538,101                  100,454    2.92% 
 grow_fnv_10_000                 419,084               349,202                    -69,882  -16.67% 
 insert_100                      4,823                 4,556                         -267   -5.54% 
 insert_1000                     46,000                45,223                        -777   -1.69% 
 insert_100_000                  6,664,378             6,213,870                 -450,508   -6.76% 
 insert_10_000                   534,694               510,209                    -24,485   -4.58% 
 insert_1_000_000                123,740,627           107,358,807            -16,381,820  -13.24% 
 insert_int_bigvalue_10_000      1,651,770             1,539,210                 -112,560   -6.81% 
 insert_str_10_000               599,655               580,417                    -19,238   -3.21% 
 insert_string_10_000            1,382,560             1,363,848                  -18,712   -1.35% 
 iter_keys_100_000               361,889               339,290                    -22,599   -6.24% 
 iter_keys_1_000_000             8,312,419             8,026,248                 -286,171   -3.44% 
 iter_keys_big_value_100_000     526,549               584,094                     57,545   10.93% 
 iter_keys_big_value_1_000_000   9,904,451             9,810,863                  -93,588   -0.94% 
 iter_values_100_000             366,842               338,926                    -27,916   -7.61% 
 iter_values_1_000_000           8,532,860             8,143,969                 -388,891   -4.56% 
 iterate_100_000                 367,828               342,287                    -25,541   -6.94% 
 iterate_1_000_000               8,542,904             8,187,025                 -355,879   -4.17% 
 lookup_100_000                  323,857               307,916                    -15,941   -4.92% 
 lookup_100_000_bigvalue         342,299               323,328                    -18,971   -5.54% 
 lookup_10_000                   262,727               251,569                    -11,158   -4.25% 
 lookup_10_000_bigvalue          282,536               264,107                    -18,429   -6.52% 
 lookup_10_000_exist             252,367               244,794                     -7,573   -3.00% 
 lookup_10_000_noexist           273,349               273,088                       -261   -0.10% 
 lookup_1_000_000                284,377               248,260                    -36,117  -12.70% 
 lookup_1_000_000_bigvalue       268,109               262,584                     -5,525   -2.06% 
 lookup_1_000_000_bigvalue_unif  660,544               687,425                     26,881    4.07% 
 lookup_1_000_000_unif           517,636               535,840                     18,204    3.52% 
 merge_shuffle                   1,488,137             1,400,964                  -87,173   -5.86% 
 merge_simple                    38,786,908            31,456,907              -7,330,001  -18.90% 
 new                             9                     9                                0    0.00% 
 with_capacity_10e5              2,500                 1,350                       -1,150  -46.00%
```

**Comment by alexcrichton** [comment]
@arthurprs hm some of those benchmarks seem worrisome indicating that iteration gets 40% slower? Is that just an outlier though?

**Comment by arthurprs** [comment]
@alexcrichton yeah, it's super odd (and makes no sense to me) but I can reproduce it every time here. I was hopping somebody else could run the benchmarks.

It's just a matter of checking out these branches (https://github.com/arthurprs/hashmap2/tree/layout and https://github.com/arthurprs/hashmap2/tree/layout_usize) and running something like
`cargo bench hhkkvv::iter --target=i686-unknown-linux-gnu`

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

Another neat improvement to our hash tables, and seems like the numbers back it up?

**Comment by alexbool** [comment]
rfcbot stuck?

**Comment by bluss** [comment]
I'm waiting for @arthurprs's hashmap change to merge first before updating. I think the benchmarks said this was a universal win for that memory layout(?), so it should be a simple decision then.

**Comment by alexcrichton** [comment]
Ah yeah we've discussed this as well, so it's ok to r+ when ready to go.

**Comment by nnethercote** [comment]
Measuring with compare.py from rustc-benchmarks would be useful here.

**Comment by alexcrichton** [comment]
@bluss this is ready to go now, right?

**Comment by bluss** [comment]
It is ready to go. I haven't invested more time in benchmarking this, and that wasn't the plan from the start either; I wanted to simply implement this and give to rustc developers if they were interested. @arthurprs's benchmarks show surprisingly noticeable gains, so I'm happy.

**Comment by alexcrichton** [comment]
@bors: r+

Ok, thanks!

**Comment by arthurprs** [comment]
The gains very likely come from smaller hash array memory footprint and cheaper displacement() calculation.

**Pull request merged**